crates/red_knot/src/symbols.rs

```diff
@@ -0,0 +1,182 @@
+#![allow(dead_code, unused_imports)]
+use hashbrown::hash_map::RawEntryMut;
+use ruff_index::{newtype_index, IndexVec};
+use ruff_python_ast::visitor::preorder::{PreorderVisitor, TraversalSignal};
+use ruff_python_ast::visitor::{preorder, Visitor};
+use ruff_python_ast::{AnyNodeRef, Expr, Mod, ModModule, Stmt, StmtExpr};
+use ruff_python_parser::{Mode, ParseError};
+use ruff_text_size::{Ranged, TextRange};
+use rustc_hash::FxHasher;
+use std::hash::{Hash, Hasher};
+
+type Map<K, V> = hashbrown::HashMap<K, V, ()>;
+
+struct SourceText {
+    text: String,
+}
+
+struct Parsed {
+    ast: ModModule,
+    imports: Vec<String>,
+    errors: Vec<ParseError>,
+}
+
+impl Parsed {
+    fn new(ast: ModModule, imports: Vec<String>, errors: Vec<ParseError>) -> Self {
+        Self {
+            ast,
+            imports,
+            errors,
+        }
+    }
+}
+
+fn parse(source: SourceText) -> Parsed {
+    let result = ruff_python_parser::parse(&source.text, Mode::Module);
+
+    let (module, errors) = match result {
+        Ok(Mod::Module(module)) => (module, vec![]),
+        Ok(Mod::Expression(expression)) => (
+            ModModule {
+                range: expression.range(),
+                body: vec![Stmt::Expr(StmtExpr {
+                    range: expression.range(),
+                    value: expression.body,
+                })],
+            },
+            vec![],
+        ),
+        Err(errors) => (
+            ModModule {
+                range: TextRange::default(),
+                body: Vec::new(),
+            },
+            vec![errors],
+        ),
+    };
+
+    Parsed::new(module, Vec::new(), errors)
+}
+
+#[newtype_index]
+struct ScopeId;
+
+#[newtype_index]
+struct SymbolId;
+
+enum ScopeKind {
+    Module,
+    Annotation,
+    Class,
+    Function,
+}
+
+struct Scope {
+    kind: ScopeKind,
+    // symbol IDs, hashed by symbol name
+    symbols_by_name: Map<SymbolId, ()>,
+}
+
+struct Symbol {
+    name: String,
```

Comment:
Nit: We should consider using `smol_str` to avoid allocating for short symbol names.